### ISSUE TITLE
panicked at 'already mutably borrowed: BorrowError' on nested signal

### ISSUE THREAD

**Issue by nguiard** [leptos-rs/leptos#1245]
**Describe the bug**
I'm getting an unexpected "panicked at 'already mutably borrowed: BorrowError'" error on the client.

**Leptos Dependencies**

```toml
leptos = { version = "0.3.1", default-features = false, features = ["serde"] }
```

**To Reproduce**

Please try to run this component on the browser. If you can't reproduce the issue, I'll try to make a full minimal repo.

```rust
#[component]
fn MinimalRepro(cx: Scope) -> impl IntoView {
    #[derive(Debug, Clone, PartialEq)]
    struct ReactiveTodo {
        id: u32,
        task: String,
        done: RwSignal<bool>,
    }

    let todos_view = create_rw_signal(cx, Some(
        vec![
            ReactiveTodo { 
                id: 0, 
                task: "Some task".to_string(),
                done: create_rw_signal(cx, false),
            },
        ]
    ));

    let toggle_todo_action = create_action(cx, move |input: &u32| {
        let input = input.clone();
        async move {
            todos_view.update(|todos| {
                let mut todos = todos.as_mut().unwrap();
                let pos = todos.iter().position(|t| t.id == input).unwrap();
                todos[pos].done.set(!(todos[pos].done)());
            });
        }
    });

    let delete_all_done_disabled = move || {
        todos_view
            .get()
            .map(|todos| todos.iter().all(|t| t.done.get() == false))
            .unwrap_or(false)
    };

    view! { cx,
        <For
            each=move || todos_view.get().unwrap_or_default()
            key=|todo| todo.id
            view=move |cx, todo| view! { cx, 
                <div
                    on:click=move |_| {
                        toggle_todo_action.dispatch(todo.id);
                    }
                >
                    { todo.clone().task }
                </div>
            }
        />
        <button disabled=delete_all_done_disabled>
            "Delete done todos"
        </button>
    }
}
```

1. Run the component browser-side
2. Open browser console
3. Click on the "Some task" text of the todo
    * you might have to click multiple times for the issue to appear
5. See error

**Expected behavior**
No panic and `toggle_todo_action` is correctly run.

**Additional context**
The rationale for what I was trying to do is : 
- I have a list of todos, render in a `<For />`, and I'm trying to be able mark them as done on click and apply some classes etc. So I've tried the path of embedding a signal in the todos for this. This works fine
- I want a button that deletes done todos on click, and *I want it to be disabled when there a no done todos*. So I'm trying to see if all todos have done == false, with the `delete_all_done_disabled` closure... This part combined with the first creates the error.

I was able to work around the issue with a `queue_microtask`, as suggested here #65 .

**Comment by gbj** [comment]
Here you update the value of a signal, inside the callback to update the value of the parent signal:
```rust
// takes out a mutable borrow of `todos_view`
todos_view.update(|todos| {
    let mut todos = todos.as_mut().unwrap();
    let pos = todos.iter().position(|t| t.id == input).unwrap();
    // tries to take out a mutable borrow of `todos[pos].done` and panics
    todos[pos].done.set(!(todos[pos].done)());
});
```

Not sure what your IDE setup is like. In mine, rust-analyzer helpfully points out that the `mut todos` is unnecessary. And indeed you're not actually mutating the list.
<img width="752" alt="Screenshot 2023-06-29 at 7 06 43 AM" src="https://github.com/leptos-rs/leptos/assets/286622/690b7ad5-4f2e-4d9a-b1ae-80190f2ad2b9">

If I replace `.update()` with its immutable-borrowing equivalent `.with()`, there is no panic.
```rust
todos_view.with(|todos| {
    let todos = todos.as_ref().unwrap();
    let pos = todos.iter().position(|t| t.id == input).unwrap();
    todos[pos].done.set(!(todos[pos].done)());
});
```

Basically, you can't nest a signal update inside a signal update. You can nest a signal update inside an access with `.with()`

Does that help?

**Comment by nguiard** [comment]
Ooh thanks that helps a lot! I understand now that at first I had Vec<Todo> without a reactive `done` field, and at that point I needed the reactive update on the Vec. After making `done` a signal, I don't need the reactive update on the Vec. 

However, I think what also threw me off is even though you say that you can't nest a signal update inside a signal update, before adding the `delete_all_done_disabled` closure, the code actually "worked"... So I was incorrectly looking for a problem in my new code and not. Regarding the warnings, I admit I was fooling around a lot and had a lot of warnings in bacon that I (badly) ignored because it was mostly unused stuff that would be used later... But I don't have rust-analyzer setup in my IDE, which would have pinpointed the problem. I think I'll do that :)

Thanks again.
